dask_expr/_core.py

```diff
@@ -28,23 +28,45 @@ def _unpack_collections(o):
 class Expr:
     _parameters = []
     _defaults = {}
+    _instances = weakref.WeakValueDictionary()
 
-    def __init__(self, *args, **kwargs):
+    def __new__(cls, *args, **kwargs):
         operands = list(args)
-        for parameter in type(self)._parameters[len(operands) :]:
+        for parameter in cls._parameters[len(operands) :]:
             try:
                 operands.append(kwargs.pop(parameter))
             except KeyError:
-                operands.append(type(self)._defaults[parameter])
+                operands.append(cls._defaults[parameter])
         assert not kwargs, kwargs
         operands = [_unpack_collections(o) for o in operands]
-        self.operands = operands
-        if self._required_attribute:
-            dep = next(iter(self.dependencies()))._meta
-            if not hasattr(dep, self._required_attribute):
+        _dependencies = [operand for operand in operands if isinstance(operand, Expr)]
```

Comment:
Agree that we shouldn't need to "attach" _dependencies to the Expr class here, but that information is being used to check `_required_attribute` below anyway.

I don't think we really *need* the `_required_attribute` check anymore if we update the `Expr.__getattr__` to avoid throwing an `AttributeError` when **any** metadata calculation (`._meta`, `._meta_chunk`, etc.) fails. I'll submit a simple PR to strip out the `_required_attribute` stuff.

Reply:
I was just lazy and don't really care what we do with `_dependencies`. Personally, I consider the cached property an indirection and a little more difficult to grasp but I don't have strong opinions here. We have to compute it either way at this point so it's just a matter of code organization.

Getting rid of the `_required_attribute` check would of course simplify this quite a bit

Reply:
This PR removes `_required_atttribute`: https://github.com/dask-contrib/dask-expr/pull/799

(not that removing that attribute solves anything here - just nice to remove unnecessary code)